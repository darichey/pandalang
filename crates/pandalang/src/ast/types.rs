// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
