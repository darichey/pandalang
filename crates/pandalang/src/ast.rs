#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}
