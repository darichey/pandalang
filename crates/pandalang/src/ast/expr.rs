#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
