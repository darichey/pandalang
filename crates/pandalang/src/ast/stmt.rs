use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
