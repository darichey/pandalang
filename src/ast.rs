#[derive(PartialEq, Eq, Debug)]
pub enum Expr {
    Int(i64),
    Var(String),
    BinOp {
        left: Box<Expr>,
        right: Box<Expr>,
        kind: BinOpKind,
    },
    Let {
        name: String,
        value: Box<Expr>,
        body: Box<Expr>,
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}
