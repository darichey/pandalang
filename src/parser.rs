use lalrpop_util::{lexer::Token, ParseError};

use crate::ast::Expr;

lalrpop_mod!(pub grammar);

pub fn parse<'input>(
    s: &'input str,
) -> Result<Box<Expr>, ParseError<usize, Token<'input>, &'static str>> {
    grammar::ExprParser::new().parse(s)
}

#[cfg(test)]
mod tests {
    use crate::{
        ast::{BinOpKind, Expr},
        parser::parse,
    };

    #[test]
    fn parses_ints() {
        assert_eq!(parse("0").unwrap(), Box::new(Expr::Int(0)));
        assert_eq!(parse("10").unwrap(), Box::new(Expr::Int(10)));
        assert_eq!(parse("-37").unwrap(), Box::new(Expr::Int(-37)));
        assert_eq!(parse("1337").unwrap(), Box::new(Expr::Int(1337)));
        assert_eq!(parse("-0").unwrap(), Box::new(Expr::Int(-0)));
    }

    #[test]
    fn parses_vars() {
        assert_eq!(parse("x").unwrap(), Box::new(Expr::Var("x".to_string())));
        assert_eq!(
            parse("foo").unwrap(),
            Box::new(Expr::Var("foo".to_string()))
        );
        assert_eq!(parse("x'").unwrap(), Box::new(Expr::Var("x'".to_string())));
    }

    #[test]
    fn parses_bin_ops() {
        assert_eq!(
            parse("1 + 1").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Int(1)),
                right: Box::new(Expr::Int(1)),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("1 - 1").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Int(1)),
                right: Box::new(Expr::Int(1)),
                kind: BinOpKind::Sub
            })
        );

        assert_eq!(
            parse("1 * 1").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Int(1)),
                right: Box::new(Expr::Int(1)),
                kind: BinOpKind::Mul
            })
        );

        assert_eq!(
            parse("1 / 1").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Int(1)),
                right: Box::new(Expr::Int(1)),
                kind: BinOpKind::Div
            })
        );

        assert_eq!(
            parse("a + b").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Var("a".to_string())),
                right: Box::new(Expr::Var("b".to_string())),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("a + b + c").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::Var("b".to_string())),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::Var("c".to_string())),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("a + b - c").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::Var("b".to_string())),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::Var("c".to_string())),
                kind: BinOpKind::Sub
            })
        );

        assert_eq!(
            parse("a - b + c").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::Var("b".to_string())),
                    kind: BinOpKind::Sub
                }),
                right: Box::new(Expr::Var("c".to_string())),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("a * b / c").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::Var("b".to_string())),
                    kind: BinOpKind::Mul
                }),
                right: Box::new(Expr::Var("c".to_string())),
                kind: BinOpKind::Div
            })
        );

        assert_eq!(
            parse("a + b * c - d / e").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::BinOp {
                        left: Box::new(Expr::Var("b".to_string())),
                        right: Box::new(Expr::Var("c".to_string())),
                        kind: BinOpKind::Mul
                    }),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("d".to_string())),
                    right: Box::new(Expr::Var("e".to_string())),
                    kind: BinOpKind::Div
                }),
                kind: BinOpKind::Sub
            })
        );
    }

    #[test]
    fn parses_parens() {
        assert_eq!(parse("(0)").unwrap(), Box::new(Expr::Int(0)));
        assert_eq!(parse("(x)").unwrap(), Box::new(Expr::Var("x".to_string())));

        assert_eq!(
            parse("(a + b)").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::Var("a".to_string())),
                right: Box::new(Expr::Var("b".to_string())),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("(a + b) + c").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::Var("b".to_string())),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::Var("c".to_string())),
                kind: BinOpKind::Add
            })
        );

        assert_eq!(
            parse("a + (b * c) - (d / e)").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::BinOp {
                        left: Box::new(Expr::Var("b".to_string())),
                        right: Box::new(Expr::Var("c".to_string())),
                        kind: BinOpKind::Mul
                    }),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("d".to_string())),
                    right: Box::new(Expr::Var("e".to_string())),
                    kind: BinOpKind::Div
                }),
                kind: BinOpKind::Sub
            })
        );

        assert_eq!(
            parse("(a + b * (c - d)) / e").unwrap(),
            Box::new(Expr::BinOp {
                left: Box::new(Expr::BinOp {
                    left: Box::new(Expr::Var("a".to_string())),
                    right: Box::new(Expr::BinOp {
                        left: Box::new(Expr::Var("b".to_string())),
                        right: Box::new(Expr::BinOp {
                            left: Box::new(Expr::Var("c".to_string())),
                            right: Box::new(Expr::Var("d".to_string())),
                            kind: BinOpKind::Sub
                        }),
                        kind: BinOpKind::Mul
                    }),
                    kind: BinOpKind::Add
                }),
                right: Box::new(Expr::Var("e".to_string())),
                kind: BinOpKind::Div
            })
        );
    }
}
