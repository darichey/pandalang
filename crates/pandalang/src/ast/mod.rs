use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
