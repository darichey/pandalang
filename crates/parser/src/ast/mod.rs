use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
