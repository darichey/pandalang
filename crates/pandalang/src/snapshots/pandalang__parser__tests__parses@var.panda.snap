---
source: src/parser.rs
expression: source
input_file: src/snapshot_inputs/var/var.panda
---
Var(
    Var {
        name: "x",
    },
)
